### PR TITLE
Minor fix for recursion into OneDrive - Change FindFirstFileEx() to use SafeFindHandle type

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

As [reported](https://github.com/PowerShell/PowerShell/pull/9509#pullrequestreview-277963361) by @KalleOlaviNiemitalo `FindFirstFileEx()` requires `FindClose()` instead of `CloseHandle()`.  Fix is to have `SafeFindHandle` class to call `FindClose()`.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 1]
```diff
@@ -7859,8 +7859,22 @@ private struct REPARSE_GUID_DATA_BUFFER
...
+            }
+
+            [DllImport(PinvokeDllNames.FindCloseDllName)]
+            [return: MarshalAs(UnmanagedType.Bool)]
+            private static extern bool FindClose(IntPtr handle);
+        }
+
         [DllImport(PinvokeDllNames.FindFirstFileDllName, EntryPoint = "FindFirstFileExW", SetLastError = true, CharSet = CharSet.Unicode)]
-        private static extern SafeFileHandle FindFirstFileEx(string lpFileName, FINDEX_INFO_LEVELS fInfoLevelId, ref WIN32_FIND_DATA lpFindFileData, FINDEX_SEARCH_OPS fSearchOp, IntPtr lpSearchFilter, int dwAdditionalFlags);
+        private static extern SafeFindHandle FindFirstFileEx(string lpFileName, FINDEX_INFO_LEVELS fInfoLevelId, ref WIN32_FIND_DATA lpFindFileData, FINDEX_SEARCH_OPS fSearchOp, IntPtr lpSearchFilter, int dwAdditionalFlags);
```
Comment (anmenaga) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: I would add text from PR description as a comment here. Otherwhise looks OK,.

**Review by anmenaga — APPROVED** [review]

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs (hunk 1 above)
Comment (SteveL-MSFT) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs]: I think the return type makes it self documenting

Comment (iSazonov) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs]: > SetLastError = true

It seems we don't use this.

Comment (KalleOlaviNiemitalo) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs]: I'd keep the SetLastError anyway, to reduce the risk of bugs if the declaration is later copied elsewhere. Or does PowerShell generally have a policy of removing unused DllImportAttribute.SetLastError for perf reasons?

Comment (iSazonov) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs]: > for perf reasons

It was my thought but I don't measure.

**Review by adityapatwardhan — APPROVED** [review]

**Pull request merged**